[1단계 - 상품 관리 기능] 로건(정다빈) 미션 제출합니다.

70825 opened the pull request:
안녕하세요 서브웨이! 😀 로건이라고 합니다 잘부탁드립니다!

---

### 테스트 코드에서 어디까지 값을 확인해야 하는지에 대한 기준?

레벨 2 미션을 진행하면서 도메인 객체까지 추가되면 굉장히 많은 테스트 코드가 작성되어야 함을 느꼈습니다. 그리고 어느 순간부터 `테스트 코드도 유지보수 대상인데, 어디까지 예상값과 실제값을 비교해야하지?`라는 생각도 같이 들었습니다.
일단은 단위 테스트를 하는 것처럼 다 확인하는 코드를 추가했습니다.

- E2E 테스트를 할 때, 예상값과 실제값을 하나하나 확인하는게 맞을까요? 그러면 유지보수 비용이 너무 높아지는 것 같은데, HTTP 상태 코드만 확인하거나, 여러개의 데이터가 들어오면 size까지 확인하는 것도 충분할까요?
- `Controller`, `Service`, `DAO`의 단위 테스트도 도메인 객체 단위 테스트처럼 예상 값과 실제 값을 하나하나 비교하는게 좋을까요?

---

### Mapper 패키지를 두고 관리

DTO에 정적 팩토리 메서드를 사용하다가 많은 정팩메가 생기길래 Mapper 패키지를 따로 두고 관리했습니다.
Mapper 클래스는 dto <-> dto로 변환할 때만 사용합니다. 그런데 이걸 따로 만들어서 관리하는게 좋은 방법인지 궁금합니다.
이전 미션에서는 정팩메가 한두개뿐이라 Mapper 클래스를 만들어서 관리할 대상을 늘리는건 굳이?라는 생각도 들었지만, 지금처럼 정팩메가 점점 생긴다면 따로 관리하는 것도 좋아보여서요.

---

뭔가 정답이 없을법한 질문들을 물어본 것 같네요 ㅋㅋ.. 서브웨이만의 기준이 있는지, 아니면 국룰이 있는지 궁금합니다.
좋은 학습 자료나 키워드 알려주시면 감사드리겠습니다!

감사합니다!

hectick committed:
docs: 기능 목록 작성

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
docs: 테이블 구조 작성

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 테이블 구조 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 엔티티 클래스 생성

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
build: mockito 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: DB 에서 상품 전체를 조회하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: DB 에서 상품 전체를 조회하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 dto 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 전체를 조회하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 전체를 조회하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 테스트 이름 추가, 테스트 목적에 부합하도록 데이터 강화

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
style: 코드 포맷 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
docs: 기능 목록 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 데이터베이스에 기본 데이터 생성 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 정보가 보이도록 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 전체 목록에 대한 응답 생성

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 전체 목록에 대한 응답 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
docs: 기능 목록 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: Model 대신 ModelAndView로 변경

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지에서 상품 정보가 보이도록 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지에서 상품 전체 목록에 대한 응답 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리페이지 상품 추가 프론트 코드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 추가 요청 dto 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: DB 에서 상품을 등록하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: DB 에서 상품을 등록하는 기능 테스트, 테스트 전략 변경
Mockito -> JdbcTest

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 인터페이스에 제품 등록 기능 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: dto에 대한 정적 팩토리 메서드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품을 등록하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 상품을 등록하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 클라이언트 요청에 대한 정적 팩토리 메서드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지에서 상품을 등록에 대한 응답 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
docs: 기능 목록 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 관리자 페이지에서 상품을 등록에 대한 응답 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 관리자 페이지에서 상품 전체 목록 조회에 대한 응답 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 메인 페이지에서 상품 전체 목록 조회에 대한 응답 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: DB 에서 상품을 수정하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: DB 에서 상품을 수정하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품을 수정하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 클라이언트 요청에 대한 정적 팩토리 메서드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 상품을 수정하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: 테스트 코드 설명 문장 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지 상품 수정 프론트 코드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
fix: 아이디를 null로 설정한 코드 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품 수정 요청 dto 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지에서 상품 수정에 대한 응답 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 관리자 페이지에서 상품 수정에 대한 응답 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: 래퍼 클래스로 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: 테이블이 존재하지 않을 때만 생성하도록 쿼리 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: DB 에서 상품을 삭제하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: DB 에서 상품을 삭제하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 인터페이스에 제품 삭제 기능 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품을 삭제하는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 상품을 삭제하는 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지 상품 삭제 프론트 코드 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 관리자 페이지에서 상품 삭제에 대한 응답 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 관리자 페이지에서 상품 삭제에 대한 응답 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: 통합 테스트 경로 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
style: 코드 포맷 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
docs: 기능 목록 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
refactor: 관리자 페이지에서 상품을 수정 요청하는 경로 변경

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: Mapper Layer 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
style: 코드 포맷 수정

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
build: validation 추가

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 클라이언트 요청에 대한 검증 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
test: 클라이언트 요청에 대한 검증 기능 테스트

Co-authored-by: 70825 <hello70825@gmail.com>

hectick committed:
feat: 상품을 추가, 수정할 때 잘못된 형식을 입력하면 화면에 알림창이 뜨는 기능 구현

Co-authored-by: 70825 <hello70825@gmail.com>

70825 committed:
refactor: @Override 적용

70825 committed:
refactor: 메서드 네이밍, 파라미터 변경
update -> updateById, delete -> deleteById

70825 committed:
refactor: final 키워드 추가, 숫자 단위를 읽기 쉽게 변경

70825 committed:
feat: 상품 컨트롤러 분리

70825 committed:
refactor: 테이블 생성 쿼리문 재정렬

70825 committed:
docs: 상품 관리 API 추가

70825 committed:
refactor: assertAll 삭제

70825 committed:
refactor: 사용하지 않는 정적 팩토리 메서드 삭제

Co-authored-by: hectick <ch.yeondev@gmail.com>

70825 committed:
refactor: 컨트롤러 이름 변경
MainController -> HomeController

70825 committed:
refactor: url 변수명 수정

70825 committed:
feat: 읽기 전용 어노테이션 추가

70825 committed:
test: 상품 추가 요청 DTO에 대한 단위 테스트 추가

70825 committed:
test: 상품 수정 요청 DTO에 대한 단위 테스트 추가

70825 committed:
refactor: 공통된 값 상수화

70825 committed:
refactor: 에러 응답 dto의 이름을 범용적으로 변경
ValidationExceptionResponse -> ExceptionResponse

70825 committed:
feat: 예상하지 못한 에러에 대한 핸들러 추가

70825 committed:
fix: 이미지를 등록할 때, URL 형식의 경로가 아니면 에러를 반환하도록 변경

70825 committed:
refactor: 메서드 순서 변경

70825 committed:
refactor: get 호출을 하는 컨트롤러는 WebMvcTest로 테스트하도록 변경

70825 committed:
refactor: 테스트 결과의 모든 값을 비교하는 것으로 변경

70825 committed:
fix: 테스트가 통과하지 않는 현상을 수정

70825 committed:
feat: request의 validation 제거, entity의 validation 추가

joseph415 commented:
로건 안녕하세요~!

어제 공유드린것 처럼 워크샵때문에 오늘 리뷰 진행해요! 오늘 새벽이나 내일 진행될것 같아서 토욜까지는 확인 가능하도록 하겠습니다~!

joseph415 commented on src/main/java/cart/controller/ExceptionController.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    @ExceptionHandler(Exception.class)
```
서버 내부오류로 Exception이 발생해도 클라이언트로 BAD_REQUEST가 내려갈거같은데, 이러면 클라이언트 측에서 에러를 파악하기 쉽지 않을 것 같습니다.

70825 replied:
앗 그렇네요.. 삭제했습니다!

joseph415 replied:
제가 말씀드린 의도는 MethodArgumentNotValidException 는 400 으로 서버 내부에서 디폴트로 내려가는 Exception는 500으로 변경해달라는 거였어요~ 이러면, 서버내부에서 IllegalArgumentException 발생시 400으로 내려가는건 똑같아서요ㅎㅎ 

4xx은 클라이언트측 문제이고 5xx은 서버측 문제로 정의되어있습니다

https://developer.mozilla.org/ko/docs/Web/HTTP/Status

70825 replied:
아아아 이해했습니다! 제가 댓글을 잘못 이해했었네요
감사합니다 😀

joseph415 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Void> postProducts(@Valid @RequestBody ProductCreationRequest request) {
```
request dto 내부에 validatation annotation 사용하는 곳이 없는데요. @Valid는 불필요하지 않나요?!

70825 replied:
Request DTO에 validation을 적용했다가 Entity에 validation을 적용하는 과정에서 삭제를 안한 것 같아요. 수정했습니다!

joseph415 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+        return ResponseEntity.status(HttpStatus.CREATED).build();
+    }
+
+    @PutMapping("/{productId}")
```
Put 과 Patch 차이를 아실까요?
해당 api는 Put보다 Patch의 기능을하고있는 것 같아서 한번 어떤 차이가 있는지 찾아보셔도 좋을 것 같습니다.

한번 깊게 찾아보셔도 좋아보여요~

70825 replied:
이전까지 Put은 데이터 전체를 수정할 때, Patch는 데이터를 일부분 수정할 때로 알고 있었습니다.
그래서 Put을 적용한 이유도 상품을 수정할 때, 모든 정보에 대한 데이터를 보내주어서 PutMapping을 사용했습니다.

그런데 Patch로 수정해서 적용해보니까 문제 없이 돌아가길래 찾아보았더니 Patch는 조건부로 멱등성이 보장되는데, Put처럼 모든 정보를 보내주는 지금이 멱등성이 보장되는 것이네요 😲
거기다가 Put은 상품이 없으면 저장하고, 있으면 수정하는 용도로 사용할 수 있지만, Patch는 무조건 수정할 때 사용할 수 있다는 걸 알았습니다. 그러면 여기서는 Patch를 사용하는게 더 적절하네요!
수정했습니다!

joseph415 replied:
넵! 잘 이해하신것 같네요. 모든 데이터도 부분집합의 일부분이잖아요?!
그렇게 생각하는것도 좋은데, Put은 전달해준 값으로 대체한다는 의미로 생각하셔도 좋습니다.

70825 replied:
아아 put이면 payload에 있는 데이터로 완전히 교체하고, patch는 수정하는 것 때문에 대체하는 부분 말하시는거죠? 감사합니다!

joseph415 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PutMapping("/{productId}")
+    public ResponseEntity<Void> putProducts(@PathVariable Long productId,
```
이것도 마찬가지로 request dto 내부에 validatation annotation 사용하는 곳이 없네요

70825 replied:
제거했습니다!

joseph415 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+import javax.validation.Valid;
+import java.util.List;
+
+@RequestMapping("/products")
```
확인 차 여쭤보아요~ 

아래 api들을 구현해주셨는데요. api 인터페이스를 어떤 생각을 갖고, 어떻게 설계 하신건지 적어주시면 감사하겠습니다~
혹시 restful api에 대해서 들어보셨나요?

70825 replied:
서브웨이 리뷰를 보고 주말동안 RESTful API에 대해 영상을 보았는데 어렵네요.. 😭

일단 기존에 알고 있던 지식으로 만들자면 아래와 같습니다.
- URI 가지고 어떤 정보인지 파악할 수 있어야합니다.
  -  상품들에 대한 정보는 복수로 `products`로 이름을 지었습니다.
  - 개별 상품에 대한 정보는 id를 통해 판단할 수 있으므로 `/products/{productId}`로 만들었습니다.

이후 여러 곳에서 글과 영상을 보고, 많은 고민을 해보며 다시 작성해봤는데, 그대로 사용하는게 맞아보였어요.
- REST는 긴 시간에 걸쳐 진화하는 애플리케이션을 위한 것이라는데, 진화란 새로운 기능이 추가되어도 쉽게 추가할 수 있고, 수정도 유연하게 할 수 있는 확장성이 있는 규칙이라고 생각했어요.
- 여러 상품들에 대한 컬렉션을 나타내는 URL인 `/products`는 좋은 네이밍이 된 것 같습니다.
- 마찬가지로 특정 상품에 대한 내용을 나타낼 때에는 `/products/{productId}`도 좋은 엔드포인트 네이밍이라고 생각해요. 기존에 `{id}`로 했다가 `productId`로 수정했는데, 다른 id가 들어올 수도 있으니 수정했으니까 문제 없어보였어요. (예를 들어서 `userId`)
- 2단계 미션으로 보자면 아마 `/users/{userId}/products`로 나타내어서 `특정 유저가 장바구니를 한 상품들`로 규칙성 있게 네이밍을 지을 수 있을 것 같아요. 만약 지금 미션을 넘어서 구매한 목록까지 염두해서 만든다면 `/users/{userId}/cart/products`로 해서 장바구니를 명시하는게 좋아보여요

1. 모든 상품에 대한 정보를 가지고 오는 것이므로 `GET /products`가 적절한 것 같습니다
2. 상품을 등록하는 것은 id값을 미리 알 수는 없으므로 `POST /products`가 맞는 것 같아요
3. 현재는 데이터를 수정하는 일 밖에 없으니 `PATCH /products/{productId}`로 표현하기 충분해보입니다. `/product/{productId}/edit`도 생각해볼 수 있지만, `/products/{productId}`만으로 명확하게 자원 식별이 가능하고, 수정한다는 의미는 PATCH로도 알리기에 충분해보여요
4. 상품을 삭제하는 것도 3번과 마찬가지로 `/products/{productId}`가 명확해보였어요.

joseph415 replied:
작성해 주신 것 보니, 잘 이해하고 작성해주신게 맞는 것 같네요~👍 지금은 초반이라 rest 아키텍처에 대해서 이해를 100% 하는것은 어려울 거라고 생각되는데요ㅎㅎ 시간이 지나면서 이해도가 올라가면 그때한번 다시 봐보세요!

제가 여쭤본 이유는 고민없이 작성하게 되면, 나중에 **왜** 라는 답변을 할 수 없는 상황이 많이있었거든요ㅎㅎ 그래서 여쭤봤어요~  api 작성경험이 많이 없지만, 고민은 해보셨으면 좋겠어서 질문드렸어요!

70825 replied:
제 이번 레벨 목표가 학습에 대한 속도는 평범해도, 왜 코드를 이렇게 작성했는지 물어보면 나름대로 충분한 답변을 할 수 있는 수준은 되자고 정했는데 아직 멀었네요 🥲 저도 서브웨이를 본받아서 더 노력해야겠습니다!
이번 레벨부터 끝까지 api 설계를 계속하니까 꾸준히 관련 글이나 영상을 찾아보면서 정리해봐야겠네요 감사합니다!

joseph415 replied:
답변해주신 내용보면 훌륭합니다! 화이팅입니다!

joseph415 commented on src/main/java/cart/dto/ProductCreationRequest.java:
```diff
@@ -0,0 +1,33 @@
+    private final String image;
+    private final Integer price;
+
+    @JsonCreator
```
요건 로건이 적용해주신거 보고 찾아보고 알았는데 신기하네요ㅎㅎ
나중에 롬복이란 것도 많이 사용하실 텐데 한번 찾아보시는것도 좋을 것 같아요~ 엄청 간단하면서 좋은 라이브러리 입니다!..

70825 replied:
그러면 2단계에서 사용해보겠습니다 ㅎㅎ

joseph415 commented on src/main/java/cart/entity/ProductEntity.java:
```diff
@@ -0,0 +1,55 @@
+    @URL
+    private final String image;
+
+    @Max(value = 10_000_000, message = "상품 등록은 최대 천만원까지 가능합니다.")
```
도메인 내에서는 valid 어노테이션 사용하지 않고, 직접 로직으로 풀면 어떤가 생각이 드는데요.

지금은 간단해서 상관은 없는데, validateion 로직이 복잡해진다면 해당 어노테이션들로 처리 안될 것 같아서요.
그리고, 해당 이것들로 발생하는 에러가 아마 MethodArgumentNotValidException 이걸로 아는데, MethodArgumentNotValidException에러가 도메인 validation exception 의 이유를 잘 나타내주는거 같지 않습니다.

MethodArgumentNotValidException 들어가보면 설명에도 다음과 같이 나와있는데
> Exception to be thrown when validation on an argument annotated with @Valid fails. Extends BindException as of 5.3.

BindException의 extend라고 하네요.

70825 replied:
수정했습니다! 앞으로는 직접 검증 로직을 만드는게 맞겠네요

joseph415 replied:
요건 한번 다시 생각해볼만해서 unresolve 했고 답변 따로 달아두었습니다!

joseph415 commented on src/main/java/cart/service/ProductManagementService.java:
```diff
@@ -0,0 +1,37 @@
+        return ProductDtoMapper.from(productEntities);
+    }
+
+    public void save(final ProductDto productDto) {
```
save 라는 의미는 id가 없으면 저장, 있으면 업데이트 인데 기능은 insert이네요.
add라고 네이밍 변경하는게 좋아보입니다

joseph415 commented on src/main/java/cart/service/ProductManagementService.java:
```diff
@@ -0,0 +1,37 @@
+        this.productDao = productDao;
+    }
+
+    public List<ProductDto> findAll() {
```
ProductEntity 리턴할 수 있었을텐데 왜 dto로 리턴하신건가요?

70825 replied:
Layered Architecture에 대해 공부해보니 가장 이상적인 방법이 계층간 서로 다른 dto를 사용하여 결합도를 낮추는게 이상적이라고 알고 있습니다. 이번 미션이 단순 CRUD 미션이지만 이전 미션에서는 사용해보지 않아서 학습 목적으로 한 번 적용해보았습니다!

joseph415 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+    @GetMapping
+    public ResponseEntity<List<ProductResponse>> getProducts() {
+        final List<ProductDto> products = managementService.findAll();
+        return ResponseEntity.ok(ProductResponseMapper.from(products));
```
dto를 리턴하고 다시 controller 에서 response로 다시 감싸는데 그렇게 하신 이유가 뭔가요? 그대로 dto 보내주면 되지 않나요~?

70825 replied:
이 부분도 위 답변과 같습니다

joseph415 replied:
저는 개인적으로 서비스에서 entity를 리턴하도록 하는게 좋다고 생각하는데요. response를 리턴하는것까지는 어느정도 타당하다고 생각이 드는데, 그걸 다시 response dto로 감싸는것은 과한 분리같다는 생각이 드네요ㅎㅎ

dto로 리턴해주실 거라면, response로 리턴하도록하는게 어떤가요?ㅎㅎ 한번 시도해 보셨다는 의도는 👍 드려요ㅎ

그리고 이거 관련해서 이전에 리뷰이랑 많은 이야기를 한적이 있어서 그거 공유드릴게요! 시간나실 때 한번 읽어보세요~ㅎ
https://github.com/woowacourse/jwp-racingcar/pull/71#discussion_r1166859737

엄청 내용이 긴데요ㅎㅎ; 
정리하면 다음과 같아요.
1. 많은 dto는 관리비용을 증가시킨다.
2. 컨트롤러는 도메인을 알아도 된다.
3. 서비스끼리 의존하는 형태면 도메인을 알고있는게 더 합리적이라고 생각한다.
4. 컨트롤러에서는 도메인로직이 불릴 일이 없다고 생각한다.

70825 replied:
글 재미있게 읽었습니다.
복잡한 규모의 프로젝트에서 DTO를 사용하는게 더 코스트가 낮은 상황이 발생하는데, 이때는 프레젠테이션 레이어랑 도메인 모델이 많은 차이가 있는 경우였군요
나중에 다른 크루가 DTO를 왜 적용 안했냐고 물어보면 위에 링크 보내면 좋을 것 같습니다 ㅋㅋ
그러면 response를 반환하도록 수정하겠습니다!

joseph415 commented on src/main/java/cart/mapper/ProductEntityMapper.java:
```diff
@@ -0,0 +1,14 @@
+import cart.dto.ProductDto;
+import cart.entity.ProductEntity;
+
+public class ProductEntityMapper {
```
빈으로 등록해서 사용할 수도 있겠네요

70825 replied:
Mapper를 빈으로 등록하는 이유가 있을까요?
Mapper에 정팩메 빆에 없어서 지금처럼 ProductEntityMapper.@@@로 사용해도 충분할 것 같은데, 따로 빈으로 등록해서 사용하는 이유가 있는지 궁금합니다

joseph415 replied:
아ㅎㅎ 바꿔달라는게 아니라, 빈으로도 등록해서 사용할 수 있겠다~ 이정도의 리뷰였어요😅 저도 로건 생각이랑 동일해요ㅎㅎ.
utils 성으로 static으로 사용해도 무방하다고 생각합니다~

70825 replied:
이걸 빈으로 등록해서 사용할 수도 있는 이유가 유틸 클래스는 객체지향에 어긋나는 클래스 때문이라서 그런 것인가요?
이 부분 말고는 굳이 빈으로 관리할 필요가 있는지 궁금합니다

joseph415 replied:
음..그건 아니고, 2가지 생각이 있었는데요.
1. 따로 상태를 들고 있지 않고있으니, 빈으로 등록할 수 있겠다라고 생각
2. 만약에 Mapper에서 다른 클래스를 주입받아 객체를 생성할 수 도 있을텐데, 그때는 빈으로서 등록해서 스프링에게 객체의 라이프사이클의 책임을 넘겨서 관리할 수 면 좋겠다 라고 생각

이렇게 2가지를 생각해서 말씀드린거고, static을 사용했을 때, 객체지향에 어긋나는 클래스라는 것은 product 처럼 어떤 상태를 갖는 도메인객체에서 적용이 되는 말이라고 생각해서 utils성 클레스에서는 맞지 않은 이야기라고 생각합니다!..

70825 replied:
아하 감사합니다! 1번의 경우엔 한번도 생각을 해보지 못했는데, 앞으로 다른 클래스 있을 때 적용해보면 좋을 것 같네요
경험삼아 2단계 진행하면서 수정해보겠습니다!

joseph415 commented on src/test/java/cart/controller/ProductControllerTest.java:
```diff
@@ -0,0 +1,185 @@
+import static org.hamcrest.Matchers.is;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class ProductControllerTest {
```
ProductIntegrationTest 에 있는 메서드 여길로 와야할것같습니다~

70825 replied:
ProductIntegrationTest는 이미 ProductControllerTest에서 확인하므로 삭제했습니다!

joseph415 commented on src/test/java/cart/controller/ProductControllerTest.java:
```diff
@@ -0,0 +1,185 @@
+import static org.hamcrest.Matchers.is;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class ProductControllerTest {
```
그냥 질문인데, 왜 해당 컨트롤러만 요것만 통합테스트로 작성해주셨나요?

joseph415 replied:
그리고 ProductControllerTest 에 사후 처리 해줘야 할 것 같습니다. 각 테스트들이 영향을 받고 있어서 순서가 달라지면 테스트가 실패할  가능성이 있어서요.

ex) add test한후에 select하면 영향을 줌

70825 replied:
> 그냥 질문인데, 왜 해당 컨트롤러만 요것만 통합테스트로 작성해주셨나요?

Product Controller는 단위 테스트를 추가로 만들 필요 없이 통합 테스트로도 충분히 테스트 할 수 있어 보였습니다.

> 그리고 ProductControllerTest 에 사후 처리 해줘야 할 것 같습니다. 각 테스트들이 영향을 받고 있어서 순서가 달라지면 테스트가 실패할 가능성이 있어서요.

이걸 삭제랑 수정을 할 때 영향이 있다는걸 금요일 아침에 발견해서, 캠퍼스 가기 전에 부랴부랴 수정하고 캠퍼스 갔다가 잊어버렸네요.. 수정했습니다!

joseph415 replied:
넵 답변감사합니다! 👍

joseph415 reviewed (changes requested):
안녕하세요. 로건

리뷰어 서브웨이입니다. 1단계 구현 잘해주셨네요👍
몇가지 리뷰 남겼습니다. 확인 부탁드릴게요🙇‍♂️

궁금한 사항은 DM 또는 리뷰 코멘트 달아주세요! 확인이 완료된 리뷰는 resolve conversation 눌러주시면 감사하겠습니다~!

joseph415 commented:
## 답변

상태코드 정보만 확인한다면, 테스트 의미가 퇴색될 것 같습니다. 
테스트라는것이 개발자에게 빠른 피드백을 주는것이 가장 큰 장점이라고 생각하는데, 버그로 인해 잘못된 값이 나와도 상태코드는 200으로 정상 처리 되어서 테스트하는 의미가 없어질 것 같아요. e2e 테스트는 처음과끝까지 모든기능이 정상적으로 동작하는지 확인하는것이잖아요~?ㅎㅎ

> Controller, Service, DAO의 단위 테스트도 도메인 객체 단위 테스트처럼 예상 값과 실제 값을 하나하나 비교하는게 좋을까요?

테스트에 대한 제 기준은 도메인, 서비스 영역 테스트는 꼼꼼하게 작성해야 한다고 생각합니다.
이 둘의 영역이 비지니스 로직을 품고 있어서요. 나머지는 일부분만 비교해도 좋다고 생각합니다.

그외에는 테스트하려는 범위, 대상을 잘 생각해보시면 어느정도 틀을 잡으실 수 있을겁니다. 

---

변환하는 메서드가 얼마 없다면 private 메서드로 처리할 수도 있을 것 같지만, 저는 Mapper 따로 만들어도 좋다고 생각합니다.
책임분리 관점에서, 두어도 무방하다고 생각해요ㅎ 그리고 dto관련해서는 리뷰로 남겨서 거기에 답변 부탁드릴게요~

70825 committed:
fix: 테스트 메서드가 이전 테스트의 영향을 받는 현상 수정
@AutoConfigureTestDatabase로 테스트 db 변경
TRUNCATE로 테이블 비우기
매 메서드마다 ID 값을 1로 다시 시작하도록 변경

70825 committed:
refactor: Exception.class를 처리하는 핸들러 삭제

70825 committed:
refactor: 사용하지 않는 메서드 삭제

70825 committed:
refactor: 컨트롤러에 validation 제거

70825 committed:
refactor: 메서드 네이밍 수정
save -> add

70825 committed:
refactor: 테스트 데이터 초기화

70825 committed:
refactor: put -> patch로 변경
지금은 없는 데이터를 생성하지 않고, 상품을 수정하기만 하므로 patch가 더 적절함

70825 committed:
refactdor: 이미 테스트하는 클래스 삭제

70825 committed:
refactor: entity에 validation 제거, 상품 이름, 이미지, 가격 객체 추가

70825 committed:
refactor: Response 코드를 간결하게 변경

70825 committed:
refactor: 객체 이름 변경
Image -> ImageUrl

70825 commented:
안녕하세요 서브웨이!
Show Resolved랑 Hide Resolved는 Outdated가 되면 다른 사람들이 보기엔 자동으로 바뀌는줄 알았는데 아니였네요 😅 이제라도 알아서 다행이에요

---

### Spring Validation은 어디에서 주로 사용하나요?

저는 도메인 객체에 사용할 수 없으면 Request DTO에만 사용할 수 있다는 생각이 들었습니다.
왜냐하면 Request DTO → Controller → DTO → Service → Entity → DAO로 계층 이동을 할텐데, Request DTO에서 검증을 해야 여러 Service 계층에서 로직을 수행할 수 있는 것 같았습니다.
이 생각이 맞을까요?

joseph415 commented:
제가 도메인에서는 Spring Validation 말고 로직으로 처리하는게 어떤지 말씀드렸는데요. Spring validation으로 좀 더 고도화해서 직접 validator를 만들어서 aop로 동작하게끔 할 수 있는 것 같습니다.

Spring validation 문서를 보면,
> There are pros and cons for considering validation as business logic, and Spring offers a design for validation (and data binding) that does not exclude either one of them. Specifically validation should not be tied to the web tier, should be easy to localize and it should be possible to plug in any validator available. Considering the above, Spring has come up with a Validator interface that is both basic ands eminently usable in every layer of an application.

이란 말이 나오는데요. 데이터 바인딩 이외에도 어느 계층에 국한되지 않도록 설계한 의도가 있는것 같아서 리뷰를 정정해야할 것 같습니다!.. 저도 validator를 직접 구현한다던가 이런 부분까지는 알고있지 못했네요ㅎㅎ..

근데, 이렇게 고도화한 수준까지 사용하지 않는거라면, dto 단에서 사용할 수 있지 않을 까 생각이 드네요ㅎㅎ 결론은 따로 제한된 것은 없는것 같습니다!..

- https://docs.spring.io/spring-framework/docs/3.0.x/reference/validation.html
- https://medium.com/@gaemi/java-%EC%99%80-spring-%EC%9D%98-validation-b5191a113f5c

70825 committed:
feat: 예상하지 못한 오류가 발생할 경우 상태코드 500을 반환한다

70825 committed:
refactor: 적용되지 않는 어노테이션 삭제

joseph415 reviewed (changes requested):
안녕하세요. 로건

일이 있어서 코멘트에 답변만 달아두었고, 추후에 chagne request를 드리려고 했는데, 이미 답변확인해주셨네요ㅎㅎ;

변경해주신 커밋들에는 따로 커맨트 달게 없는 것 같아서, 추가 답변만 한번 확인해주세요~ 
확인 후에, 다시 리뷰신청눌러주시면 이번 단계는 merge하겠습니다~

70825 committed:
refactor: ProductDto 삭제
Service에서는 파라미터를 Entity로 받고, 반환을 Response로 하도록 변경

70825 commented:
안녕하세요 서브웨이! 덕분에 알차게 학습한 것 같아서 너무 좋았습니다!
Validation을 직접 만들어서 사용할 수 있는건 처음 알았네요. 공식 문서를 통해 공부한다는게 이런 의미군요..! 미디움 링크는 두고두고 볼 수 있는 링크라 잘 사용하겠습니다 ㅋㅋㅋ
2단계도 잘 부탁드립니다~!!

joseph415 reviewed (approved):
안녕하세요 로건!

1단계는 이만 merge하겠습니다!
다음 단계도 화이팅입니다!💯
